### PR TITLE
GPA layer and NLL loss functiona added

### LINKED ISSUE #71: Add Gaussian Process Approximation (GPA) layer
We are using GPA for quantifying uncertainty in many of our projects. It is better to have a version of GP layer in the core so far anyone can import that from the core if needed.

- [x] Create an issue and branch
- [x] Add GP layer and Negative log likelihood loss function
- [x] Add unit test
- [x] Create a Pull request

### PR DESCRIPTION
**Pull request by hasan2m**
Unit test passed for both GPA layer and loss function. 

GPA layer: 

image

Loss function:

image

### CODE REVIEW

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            trainable=False,
+            name='Lambda'
+        )
+
+       self.noise_scale = self.add_weight(
+            shape=(),  # Scalar variable
+            initializer=tf.constant_initializer(self.initial_noise_scale),
+            trainable=self.train_noise_scale,
+            dtype=tf.float32,
+            constraint=ClipByValue(1e-6, 1e6),
```
Comment (sgoldenCS) [inline jlab_datascience_toolkit/utils/keras_layers/GP_layer.py]: I like the use of a function for the constraint, but would it be possible to continue using the `noise_bounds` parameter here? Otherwise it should be removed from the class definition as I don't believe it's used elsewhere.

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        )
+
+
+        # self.noise_scale = tf.Variable(
+        #     self.initial_noise_scale,
+        #     dtype=tf.float32,
+        #     trainable=self.train_noise_scale,
+        #     constraint=lambda z: tf.clip_by_value(z, self.noise_bounds[0], self.noise_bounds[1]),
+        #     name='noise_scale'
+        # )
```
Comment (sgoldenCS) [inline jlab_datascience_toolkit/utils/keras_layers/GP_layer.py]: This looks to be essentially the same as the new version (other than the `noise_bounds` thing), so feel free to remove the commented section to clean things up.

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    #             self.update_cov(self.prior)
+
+    #         variances = self.calc_variance(ffs)
+
+    #     stddevs = tf.math.sqrt(variances)
+    #     out = [output, stddevs[:, None]]
+    #     if return_features:
+    #         out.append(ffs)
+
+    #     return out
```
Comment (sgoldenCS) [inline jlab_datascience_toolkit/utils/keras_layers/GP_layer.py]: I don't see any changes here other than the default for the `training` flag, so I think this commented section could be removed too.

**Review by sgoldenCS — COMMENTED** [review]
I generally think things look fine although I haven't run the unittests myself (maybe someone else could go through that process?). I'm glad you're using pytest and pytest fixtures though! It looks good!

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:1]: GaussianProcessLayer --> We should rename since this is RFF approximation and not GP

Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:1]: please change file name --> gpa_layer.py

File: jlab_datascience_toolkit/utils/keras_losses/nll_loss.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        pred (tf.Tensor): Predicted mean of the GP model (shape: [batch_size, 1]).
+        y (tf.Tensor): Ground truth target values (shape: [batch_size, 1]).
+        std (tf.Tensor): Standard deviation predictions from the GP model (shape: [batch_size, 1]).
+        noise_scale (tf.Tensor or float): Noise scale (variance) of the GP model.
+
+    Returns:
+        tf.Tensor: The computed NLL loss (scalar).
+    """
+    
+    sigma_star = tf.square(std) + noise_scale + 1e-5  # Adding a small constant for numerical stability
```
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_losses/nll_loss.py]: is the noise scale additive?

File: utests/utest_gp_layer.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import pytest
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import Model, Input
+from jlab_datascience_toolkit.utils.keras_layers.GP_layer import GaussianProcessLayer
```
Comment (schr476) [inline utests/utest_gp_layer.py]: update with request change above.
Please don't call it a GaussianProcessLayer.

File: utests/utest_gp_layer.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+def test_layer_instantiation():
+    try:
+        layer = GaussianProcessLayer()
+        assert layer is not None, "Failed to instantiate GaussianProcessLayer."
+    except Exception as e:
+        pytest.fail(f"Layer instantiation failed: {e}")
+
+
+def test_forward_pass(random_input):
+    layer = GaussianProcessLayer()
```
Comment (schr476) [inline utests/utest_gp_layer.py]: Please don't call it GaussianProcessLayer

**Review by schr476 — CHANGES_REQUESTED** [review]

File: jlab_datascience_toolkit/utils/keras_losses/nll_loss.py (hunk 4 above)
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_losses/nll_loss.py]: Removing the noise_scale from here because we are accounting the noise scale during the variance calculation within the gpa_layer class.

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:1]: Name changed to GaussianProcessApproxiamtionLayer.

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py (hunk 1 above)
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_layers/GP_layer.py]: Removed noise_bound from the class definition.

**Comment by hasan2m** [comment]
Addressed the comments and reran the unit test. 

image

image

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py
Comment (schr476) [reply jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:1]: remove this from utils.

File: jlab_datascience_toolkit/utils/keras_layers/GP_layer.py (hunk 3 above)
Comment (schr476) [reply jlab_datascience_toolkit/utils/keras_layers/GP_layer.py]: clean up code.

**Comment by Kishanrajput** [comment]
@hasan2m - could you please move the layer and loss out from utils by following below directory structure
Toolkit
 - keras (dir)
   - layers
     - layer_v0.py (implementation)  
   - losses (dir)
     - loss_v0.py (implementation)

**Review by schr476 — CHANGES_REQUESTED** [review]

File: jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            ffs = tf.math.sqrt(2.0 / self.n_fourier_features) * ffs
+
+        ffs = tf.math.sqrt(self.constant_scale) * ffs
+        output = self.rff_output(ffs)
+
+        batch_size = tf.cast(tf.shape(inputs)[0], tf.float32)
+
+        if training:
+            update_prior_op = (
+                self.momentum * self.prior + (1 - self.momentum) * (tf.transpose(ffs) @ ffs / batch_size)
```
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py]: check of the size is correct: (batch,batch)

File: jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+    
+    def update_cov(self, prior):
+        eigvals, eigvecs = tf.linalg.eigh(prior)  #EigenDecomposition
+        eigvals = tf.where(eigvals > 0, eigvals, tf.zeros_like(eigvals))
+
+        self.eigvals.assign(eigvals)
+        self.eigvecs.assign(eigvecs)
+
+    def calc_variance(self, ffs):
```
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py]: need to understand this better. this should be K_xx^{-1}

File: jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        
+       self.rff_map = layers.Dense(
+            self.n_fourier_features // 2,
+            trainable=False,
+            kernel_initializer=initializers.RandomNormal(mean=0.0, stddev=1.0),
+            bias_initializer=initializers.RandomUniform(0, 2 * np.pi),
+            name='rff_map'
+        )
+        
+       self.rff_output = layers.Dense(self.n_out, use_bias=False, name='GP_mean_pred')
```
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py]: What is this here?

File: jlab_datascience_toolkit/utils/keras_losses/nll_loss.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        pred (tf.Tensor): Predicted mean of the GP model (shape: [batch_size, 1]).
+        y (tf.Tensor): Ground truth target values (shape: [batch_size, 1]).
+        std (tf.Tensor): Standard deviation predictions from the GP model (shape: [batch_size, 1]).
+        
+
+    Returns:
+        tf.Tensor: The computed NLL loss (scalar).
+    """
+    
+    sigma_star = tf.square(std) + 1e-5  # Adding a small constant for numerical stability
```
Comment (schr476) [inline jlab_datascience_toolkit/utils/keras_losses/nll_loss.py]: sigma_star should be sigma2_star

**Review by schr476 — CHANGES_REQUESTED** [review]

File: jlab_datascience_toolkit/utils/keras_losses/nll_loss.py (hunk 10 above)
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_losses/nll_loss.py]: corrected.

File: jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py (hunk 8 above)
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py]: Inverse of K_xx is determined using eigenvalue and eigenvectors.

**Comment by schr476** [comment]
@hasan2m what is the status of the requested changes?

File: jlab_datascience_toolkit/keras/layers/gpa_layer_v0.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+        return result
+
+    
+    def update_cov(self, prior):
+        eigvals, eigvecs = tf.linalg.eigh(prior)  #EigenDecomposition
+        eigvals = tf.where(eigvals > 0, eigvals, tf.zeros_like(eigvals))
+
+        self.eigvals.assign(eigvals)
+        self.eigvecs.assign(eigvecs)
```
Comment (sgoldenCS) [inline jlab_datascience_toolkit/keras/layers/gpa_layer_v0.py:162]: I wonder if this should be an internal version (`_update_cov(self, prior)` or something similar) that can be called with any prior, and implement the `update_cov` function without the extra input parameter like this:
```python
def update_cov(self):
    self._update_cov(self.prior)
```
That could make it easier for a user to update the variance calculations outside of the model. Going along with this, it might also make sense to just remove the `update_cov()` call on line 146, and just rely on the user to properly update the prior when desired. At the moment, if `do_custom_cov_update == False`, the code will be slow for every non-training/validation call, which probably isn't desirable.

File: jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py (hunk 9 above)
Comment (hasan2m) [reply jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py]: Output layer that provides the mean prediction using the fourier features created by rff_map.

**Review by Kishanrajput — CHANGES_REQUESTED** [review]
When a model that uses this layer is saved and loaded back - it does not save/load prior matrix which makes the saved models not useful. Please add prior saving and loading.

**Comment by Kishanrajput** [comment]
@hasan2m - please have a look at the comments and address/comment on them.

**Comment by Kishanrajput** [comment]
Print statement for the RFF shape should be removed
